File: LeetCode/Python3/Problem 190.py
# LeetCode: Problem 190
# Jonathan Kosasih

#Reverse bits of a given 32 bits unsigned integer.

#Example 1:
#Input: 00000010100101000001111010011100
#Output: 00111001011110000010100101000000
#Explanation: The input binary string 00000010100101000001111010011100 represents the unsigned integer 43261596,
#so return 964176192 which its binary representation is 00111001011110000010100101000000.
#
#Example 2:
#Input: 11111111111111111111111111111101
#Output: 10111111111111111111111111111111
#Explanation: The input binary string 11111111111111111111111111111101 represents the unsigned integer 4294967293,
#so return 3221225471 which its binary representation is 10101111110010110010011101101001.
#------------------------------------------------------------------
# INPUT IS ACTUALLY THE INTEGER REPRESENTATION, NOT THE BINARY FORM
# RETURN IS ALSO THE INTEGER REPRESENTATION
class Solution:
    # @param n, an integer
    # @return an integer
    def reverseBits(self, n):
        # bit manipulation
        # reverse and build it up
        reverse = 0
        for i in range(32):
            reverse = reverse << 1
            reverse |= (n >> i) & 0x1
        return reverse
        
sol = Solution()
example = 4294967293
print(str(sol.reverseBits(example)))

File: LeetCode/C++/Problem 190.cpp
// LeetCode: Problem 190
// Jonathan Kosasih
/*
Reverse bits of a given 32 bits unsigned integer.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

class Solution {
public:
	uint32_t reverseBits(uint32_t n) {
		int i = 0, j = 31;
		while (i < j) {
			if (((n >> i) & 1) != ((n >> j) & 1)) {
				n ^= ((1U << i) | (1U << j));
			}
			i++, j--;
		}
		return n;
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}